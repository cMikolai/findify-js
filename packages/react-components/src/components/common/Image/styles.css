/*
.root{
  will-change: filter, opacity, height;
  transform: filter .3s linear, opacity .3s linear, height .1s linear;
  display: none;
}

.thumbnail{
  display: block;
  height: 150px;
  opacity: .5;
  filter: blur(10px);
}

.original{
  display: block;
  opacity: 1;
  filter: blur(0);
  width: 100%;

}
*/
.root {
  will-change: filter, opacity, height;
  transform: filter .3s linear, opacity .3s linear, height .1s linear;
  width: 100%;
  position: relative;

  img {
    display: block;
    width: 100%;
    height: auto;
  }
}

.thumbnail {
  opacity: .5;
  filter: blur(10px);
}

.croppedRoot {
  will-change: filter, opacity, height;
  transform: filter .3s linear, opacity .3s linear, height .1s linear;
  display: block;
  width: 100%;
  position: relative;
}

.loading {
  background-color: #F2F4F7
}
