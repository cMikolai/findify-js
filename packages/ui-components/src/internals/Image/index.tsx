import * as React from 'react';
import {
  compose,
  lifecycle,
  onlyUpdateForKeys,
  pure,
  withProps,
  mapProps,
  withStateHandlers,
  withPropsOnChange,
} from 'recompose';
import * as cx from 'classnames';
import sizeMe from 'react-sizeme';

const styles = require('./styles.css');

// const isHDPI = (() =>
//   ((window.matchMedia && (window.matchMedia('only screen and (min-resolution: 124dpi), only screen and (min-resolution: 1.3dppx), only screen and (min-resolution: 48.8dpcm)').matches || window.matchMedia('only screen and (-webkit-min-device-pixel-ratio: 1.3), only screen and (-o-min-device-pixel-ratio: 2.6/2), only screen and (min--moz-device-pixel-ratio: 1.3), only screen and (min-device-pixel-ratio: 1.3)').matches)) || (window.devicePixelRatio && window.devicePixelRatio > 1.3))
// )();

function round(value, exp) {
  value = +value;
  exp = +exp;
  value = value.toString().split('e');
  value = Math.ceil(+(value[0] + 'e' + (value[1] ? +value[1] - exp : -exp)));
  value = value.toString().split('e');
  return +(value[0] + 'e' + (value[1] ? +value[1] + exp : exp));
}

export interface OwnProps {
  src: string;
  alt?: string;
  className?: string;
  onClick?: React.MouseEventHandler<HTMLImageElement>;
}

/**
 * STAGE:
 * 0 - nothing loaded
 * 1 - thumbnail loaded
 * 2 - image loaded
*/
export interface State {
  src: string;
  stage: number;
  setSrc: (isMounted: boolean) => void;
  setThumbnail: (isLoading: boolean) => void;
}

export type MappedProps = OwnProps & { src?: string };
export type Props = MappedProps & State;

const fetch = (src: string) =>
  new Promise(resolve => {
    const img = new Image();
    img.addEventListener('load', () => resolve(src), false);
    img.src = src;
  });

const getShopifyUrl = (
  src: string,
  isThumb: boolean,
  size: { width?: number }
) => {
  const resolution = isThumb ? '19x' : `${round(size.width, 1)}x`;
  return src.replace(/_(medium|large|small)./g, `_${resolution}.`);
};

const ImageComponent = compose<OwnProps, Props>(
  // sizeMe({ refreshRate: 50 }),
  withPropsOnChange(['src'], ({ src, size }) => {
    // if (src.includes('.shopify')) {
    //   return {
    //     src: void 0,
    //     original: getShopifyUrl(src, false, size),
    //     thumbnail: getShopifyUrl(src, true, size),
    //   };
    // }
    return { src: void 0, original: src };
  }),
  withStateHandlers(
    { src: void 0, stage: 0 },
    {
      setSrc: () => src => ({ src, stage: 2 }),
      setThumbnail: () => src => ({ src, stage: 1 }),
    }
  ),
  withPropsOnChange(
    ['thumbnail', 'original'],
    ({ setSrc, setThumbnail, thumbnail, original }) => {
      if (thumbnail) {
        fetch(thumbnail)
          .then(setThumbnail)
          .then(() => fetch(original))
          .then(setSrc);
      } else {
        fetch(original).then(setSrc);
      }
    }
  ),
  mapProps(({ stage, className, src, alt, onClick }: Props) => ({
    alt,
    onClick,
    src,
    className: cx(
      styles.root,
      className,
      stage === 0 && styles.loading,
      stage === 1 && styles.thumbnail,
      stage === 2 && styles.original
    ),
  })),
  onlyUpdateForKeys(['src'])
)(props => <img {...props} />);

export default ImageComponent;
