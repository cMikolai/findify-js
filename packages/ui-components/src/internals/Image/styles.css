.root{
  will-change: filter, opacity, height;
  transform: filter .3s linear, opacity .3s linear, height .1s linear;
}

.loading{
  display: none;
}
.thumbnail{
  height: 150px;
  opacity: .5;
  filter: blur(10px);
}

.original{
  opacity: 1;
  filter: blur(0);
}
