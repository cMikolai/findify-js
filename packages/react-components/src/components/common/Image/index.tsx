import React from 'react'
import Branch from 'components/common/Branch'
import classNames from 'classnames'
import sizeMe from 'react-sizeme'

import {
  compose,
  pure,
  lifecycle,
  onlyUpdateForKeys,
  withProps,
  mapProps,
  withStateHandlers,
  withPropsOnChange,
  setDisplayName
} from 'recompose';

import styles from 'components/common/Image/styles.css';

const prefetchedImages = {}

const prefetchImage = (src: string) =>
  new Promise(resolve => {
    const img = new Image();
    img.addEventListener('load', () => {
      prefetchedImages[src] = true
      resolve(src)
    }, false);
    img.src = src;
  });

/** This is a list of props which Image component accepts */
interface ImageProps {
  /** Custom classname */
  className?: string,
  /** Source to original image */
  src: string,
  /** Source to thumbnail, which will be showed blurred during loading */
  thumbnail?: string,
  /** Width / height ratio, to which image will conform */
  aspectRatio?: number,
  /** @hidden */
  size: { width: number },
  /** @hidden */
  isFixedRatio: boolean
}

export default compose(
  sizeMe({
    monitorWidth: true,
    refreshRate: 50
  }),
  setDisplayName('Image'),
  onlyUpdateForKeys(['src', 'thumbnail']),
  withPropsOnChange(['src'], ({ src, size }) => {
    return { src: prefetchedImages[src] ? src : void 0, original: src, }
  }),
  withStateHandlers(
    ({ src, original }) => ({ src, stage: src === original ? 2 : 0 }),
    {
      setSrc: () => src => ({ src, stage: 2 }),
      setThumbnail: () => src => ({ src, stage: 1 })
    }
  ),
  withPropsOnChange(
    ['thumbnail', 'original'],
    ({ setSrc, setThumbnail, thumbnail, original, src, stage }) => {
      if (stage === 2) return;
      if (thumbnail) {
        prefetchImage(thumbnail)
          .then(setThumbnail)
          .then(() => prefetchImage(original))
          .then(setSrc);
      } else {
        prefetchImage(original).then(setSrc);
      }
    }
  ),
  withProps(({ aspectRatio }) => ({
    isFixedRatio: aspectRatio && typeof aspectRatio === 'number' && !isNaN(aspectRatio) &&isFinite(aspectRatio),
  })),
  withProps(({ className, stage, isFixedRatio }) => ({
    className: classNames(
      className,
      {
        [styles.root]: !isFixedRatio,
        [styles.croppedRoot]: isFixedRatio,
        [styles.loading]: stage === 0,
        [styles.thumbnail]: stage === 1,
        [styles.original]: stage === 2,
      }
    )
  })),
)(({ src, aspectRatio, size: { width }, className, isFixedRatio }: ImageProps) => (
  <div className={className} style={isFixedRatio ? {
    height: 1 / aspectRatio * width,
    backgroundImage: `url(${src})`,
    backgroundSize: 'cover',
  }: {}}>
    <img display-if={!isFixedRatio} src={src} />
  </div>
))
