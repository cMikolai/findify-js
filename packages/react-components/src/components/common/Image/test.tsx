import React from 'react';
import { shallow, render } from 'enzyme';
import Image from './index';

/** FIXME: although it _does_ render it correctly, this test should be expanded to show all stages of loading an image */
describe('<Image />', () => {
  it('renders correctly', () => {
    const properties = [
      { src: 'foo.png' },
      { src: 'bar.png', className: 'qux' },
    ];
    properties.forEach((props) => {
      const component = render(<Image {...props} />);
      expect(component).toMatchSnapshot();
    });
  });
});
