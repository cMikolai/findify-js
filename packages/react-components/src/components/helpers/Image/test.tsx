import React from 'react';
import { shallow, render } from 'enzyme';
import Image from './index';

describe('<Image />', () => {
  it('renders correctly', () => {
    const properties = [
      { src: 'foo.png' },
      { src: 'bar.png', className: 'qux' },
    ];
    properties.forEach((props) => {
      const component = render(<Image {...props} />);
      expect(component).toMatchSnapshot();
    });
  });
});
