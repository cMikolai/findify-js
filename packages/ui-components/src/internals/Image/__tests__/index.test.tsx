import * as React from 'react';
import { shallow, render } from 'enzyme';
import Image, { Props } from '..';

describe('<Image />', () => {
  // TODO: fix messed up context binding in Image.onload in lifecyle HOC
  it.skip('renders correctly', () => {
    const properties: Props[] = [
      { src: 'foo.png' },
      { src: 'bar.png', className: 'qux' },
    ];
    properties.forEach((props: Props) => {
      const component = render(<Image {...props} />);
      expect(component).toMatchSnapshot();
    });
  });
});
