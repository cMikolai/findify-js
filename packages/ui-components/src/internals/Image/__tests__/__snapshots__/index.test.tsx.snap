// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> renders correctly 1`] = `
<img
  class="loading"
/>
`;

exports[`<Image /> renders correctly 2`] = `
<img
  class="qux loading"
/>
`;
